### ISSUE TITLE
RendGuard should also account rendezvous circuits attempt that failed, not just established ones

### ISSUE THREAD

**Issue by gits7r** [mikeperry-tor/vanguards#52]
There's some noise on tor-dev@ about the well known attack that Rendguard component mitigates:

https://lists.torproject.org/pipermail/tor-dev/2020-February/014146.html

From what I've read it appears that some attackers might choose rendezvous points that are not Tor relays at all (random IP addresses from the internet with fake identity relay fingerprints). I don't understand what such an attacker can gain, maybe wait at the end of a hostile middle relay to see where is the circuit that requests to extend to the fake address comes from (?) - anyway not too important.

Bottom line is that Rendguard should also account failed rendezvous circuits (like ones with rendezvous points that are not Tor relays at all, or do not respond to the establish rendezvous request) and ban such destinations after few failed attempts. It's also healthy for the middles in the network overall, not to try too many times to extend to bogus destinations.

**Comment by gits7r** [comment]
Is the RendGuard component identifying and mapping rendezvous points by relay fingerprint, IP address or both relay fingerprint + IP address (and if this latter is the case, what happens if just one of them changes like same fingerprint but new IP or new fingerprint and same IP)?

We have these types of potential malicious rendezvous points:
- regular relays that are in the consensus and are running Tor normally and transparently;
- relays that are not in the consensus, but they are running Tor and the port is open, listening for OR connections and responding to them;
- relays with fake identity fingerprint and/or IP address (not running Tor or listening for OR connections)
- relays with reversed bits IP addresses

Which of these do we successfully filter currently?

**Comment by mikeperry-tor** [comment]
If there are rend points not in the consensus, we subject it to the `rend_use_max_consensus_weight_churn` limit, which is no more than 1% of the total consensus weight.

RendGuard currently only tracks by fingerprint, but that `rend_use_max_consensus_weight_churn` value applies to the total of *all* fingerprints not in the consensus (it assigns a special "fingerprint id" to relay fingerprints not present).

I am not 100% sure yet that all methods of rend failure will still be recorded, but I believe so. We start counting on the connection attempt, based on control port string HSSR_CONNECTING.

**Comment by mikeperry-tor** [comment]
I'm gonna flag this as "needs testing" because I *believe* s7rs concerns are already properly handled, but it would be great if someone could test+confirm.

**Comment by gits7r** [comment]
I will test this myself soon.  99% of my concerns are covered by your reply and confirmation, but there is just one thing that I don't understand, and if you have time I'd like to know so I can make the test properly and cook a Tor client to send certain INTRODUCTION requests in my test network.

I am sorry for not explaining properly. Based on your explanation, I also believe the concerns are properly handled. What I was thinking is, based on the report we had on the mail list, an attacker could flush fingerprints (generate new onion identity keys) for every N introductions sent for a malicious rendezvous relay under his control. Given the `rend_use_max_consensus_weight_churn` value applies to ALL the fingerprints not in the consensus, this model obviously gives no gain to an attacker as the total consensus weight churn will not be exceeded.

And if the relay is in the consensus, it is verified by the directory authorities so it has to listen on the advertised IP address and port, so the model does not work here as well. I was slightly confused because there is a scenario I did not quite understand:

say a relay that is in the consensus has:
fingerprint: ABCDEFG
IP:port: x.x.x.x:9001

If a client will send an INTRODUCE with the following data for rendezvous relay:
fingerprint = ABCDEFG
IP:port: y.y.y.y:9001

so same fingerprint, but different IP address, where will the onion service try to connect to? The IP in the consensus, or the IP in the INTRODUCE request (given the fingerprint is the same, just IP addresses differ). Does it prioritize the consensus, or the INTRODUCE request data?

I know the connection is technically done by the 3rd hop, not the onion service itself, but the question still stands. What IP address does the onion service sends instructions to extend to, the one in the consensus or the one it got in the INTRODUCE request. What is the expected and wanted behavior here?

**Comment by mikeperry-tor** [comment]
> I was slightly confused because there is a scenario I did not quite understand:
> 
> say a relay that is in the consensus has:
> fingerprint: ABCDEFG
> IP:port: x.x.x.x:9001
> 
> If a client will send an INTRODUCE with the following data for rendezvous relay:
> fingerprint = ABCDEFG
> IP:port: y.y.y.y:9001
> 
> so same fingerprint, but different IP address, where will the onion service try to connect to? The IP in the consensus, or the IP in the INTRODUCE request (given the fingerprint is the same, just IP addresses differ). Does it prioritize the consensus, or the INTRODUCE request data?
> 
> I know the connection is technically done by the 3rd hop, not the onion service itself, but the question still stands. What IP address does the onion service sends instructions to extend to, the one in the consensus or the one it got in the INTRODUCE request. What is the expected and wanted behavior here?

Rend requests are placed in RELAY_COMMAND_EXTEND cells. These are handled in `circuit_extend()`. The connection to use for the extend is determined by `channel_get_for_extend()`. The channel is looked up by identity key/fingerprint. If an open channel exists with the same fingerprint/identity key, it is used, so long as it is not marked bad for new circuits (too old), not marked for close/closing, and is the "canonical" connection for that fingerprint/identity key.

So, basically this means that if a usable, good connection already exists for a fingerprint listed in an extend cell, it will succeed, even if the IP address is wrong.

However, if a connection does not exist, then the IP address and port **from the extend cell** is passed to channel_connect_for_circuit(), so the connection will fail in that case.

I am not sure if this is the best possible outcome, but it is what happens today.

**Comment by mikeperry-tor** [comment]
It sounds like vanguards is doing the right thing here, so I am closing this. Reopen/re-file if you notice otherwise.
